sandbox/cgroup/monitor.go

```
@@ -0,0 +1,151 @@
+		basePath := path.Dir(fullPath) // Monitor the path containing this folder
+		_, exists := currentCGroupMonitor.watched[basePath]
+		if exists {
+			continue
+		}
+		err := currentCGroupMonitor.watcher.AddWatch(basePath, inotify.InDelete)
+		if err != nil {
+			continue
+		}
+		if _, err := os.Stat(fullPath); errors.Is(err, os.ErrNotExist) {
+			// if the file/folder to monitor doesn't exist after the parent being added, remove it
+			currentCGroupMonitor.watcher.RemoveWatch(basePath)
+			continue
+		}
+		currentCGroupMonitor.watched[basePath] = append(currentCGroupMonitor.watched[basePath], newApp)
+		addedPaths = true
+	}
+	if !addedPaths {
+		// if the files/folders to monitor don't exist, send the notification now
+		newApp.channel <- newApp.name
+	}
+}
+
+func monitorMainLoop() {
+	for {
+		select {
+		case event := <-currentCGroupMonitor.watcher.Event:
+			if event.Mask&inotify.InDelete != 0 {
+				onFilesDeleted(event.Name)
+			}
+		case newApp := <-currentCGroupMonitor.channel:
+			onFilesAdded(&newApp)
+		}
+	}
+}
+
+// MonitorSingleton launches the main loop and returns the CGroupMonitor singleton
+func MonitorSingleton() *CGroupMonitor {
+	if currentCGroupMonitor.watcher == nil {
+		wd, err := inotify.NewWatcher()
+		if err != nil {
+			return nil
+		}
+		currentCGroupMonitor.watcher = wd
+		go monitorMainLoop()
+	}
+	return &currentCGroupMonitor
+}
+
+// MonitorSnap is the method to call to monitor the running instances of an specific Snap.
+// It receives the name of the snap to monitor (for example, "firefox" or "steam")
+// and a channel. The caller can wait on the channel, and when all the instances of
+// the specific snap have ended, the name of the snap will be sent through the channel.
+// This allows to use the same channel to monitor several snaps
+func (this *CGroupMonitor) MonitorSnap(snapName string, channel chan string) {
+	paths, _ := InstancePathsOfSnap(snapName, InstancePathsFlagsOnlyPaths)
```

we almost never ignore errors, and when we do we need to explain why and still log them